core/src/main/scala/org/apache/spark/shuffle/BlockStoreShuffleReader.scala

```diff
@@ -42,6 +46,12 @@ private[spark] class BlockStoreShuffleReader[K, C](
 
   /** Read the combined key-values for this reduce task */
   override def read(): Iterator[Product2[K, C]] = {
+    val memMode =
```

Comment:
Yes, ideally this should be moved into `ShuffleBlockFetcherIterator`, but I didn't find a better implementation other than 
```
      extends MemoryConsumer(tmm, tmm.pageSizeBytes(),
        if (SparkTransportConf.fromSparkConf(SparkEnv.get.conf, "shuffle").preferDirectBufs()) {
          MemoryMode.OFF_HEAP
        } else {
          MemoryMode.ON_HEAP
        }
      )
```
And I'd be a little bit hesitant to expose a 'setMode' in `MemoryConsumer`